File: apps/ws/src/GameManager.ts
import { WebSocket } from 'ws';
import {
  GAME_OVER,
  INIT_GAME,
  JOIN_GAME,
  MOVE,
  OPPONENT_DISCONNECTED,
  JOIN_ROOM,
  GAME_JOINED,
  GAME_NOT_FOUND,
  GAME_ALERT,
  GAME_ADDED,
  GAME_ENDED,
} from './messages';
import { Game, isPromoting } from './Game';
import { db } from './db';
import { SocketManager, User } from './SocketManager';
import { Square } from 'chess.js';
import { GameStatus } from '@prisma/client';

export class GameManager {
  private games: Game[];
  private pendingGameId: string | null;
  private users: User[];

  constructor() {
    this.games = [];
    this.pendingGameId = null;
    this.users = [];
  }

  addUser(user: User) {
    this.users.push(user);
    this.addHandler(user);
  }

  removeUser(socket: WebSocket) {
    const user = this.users.find((user) => user.socket === socket);
    if (!user) {
      console.error('User not found?');
      return;
    }
    this.users = this.users.filter((user) => user.socket !== socket);
    SocketManager.getInstance().removeUser(user);
  }

  removeGame(gameId: string) {
    this.games = this.games.filter((g) => g.gameId !== gameId);
  }

  private addHandler(user: User) {
    user.socket.on('message', async (data) => {
      const message = JSON.parse(data.toString());
      let gameToJoin = null;
      if (message.type === INIT_GAME) {
        const gameTime = message.time;
        if (![1, 3, 5, 10].includes(gameTime)) {
          console.error('Invalid time amount: ' + gameTime);
          return;
        }
        const gamesWithSameTimeLimit = this.games.filter(
          (game) => game.gameTime === gameTime && game.player2UserId === null,
        );
        if (this.pendingGameId) {
          const game = this.games.find((x) => x.gameId === this.pendingGameId);
          if (!game) {
            console.error('Pending game not found?');
            return;
          }
          if (gamesWithSameTimeLimit.length > 0) {
            gameToJoin = gamesWithSameTimeLimit[0];
          }
          this.pendingGameId = null;
        } else {
          if (gamesWithSameTimeLimit.length > 0) {
            gameToJoin = gamesWithSameTimeLimit[0];
          } else {
            const game = new Game(user.userId, null, gameTime);
            this.games.push(game);
            this.pendingGameId = game.gameId;
            SocketManager.getInstance().addUser(user, game.gameId);
            SocketManager.getInstance().broadcast(
              game.gameId,
              JSON.stringify({
                type: GAME_ADDED,
              }),
            );
            return;
          }
        }
        if (gameToJoin) {
          if (user.userId === gameToJoin.player1UserId) {
            SocketManager.getInstance().broadcast(
              gameToJoin.gameId,
              JSON.stringify({
                type: GAME_ALERT,
                payload: {
                  message: 'Trying to Connect with yourself?',
                },
              }),
            );
            return;
          }
          SocketManager.getInstance().addUser(user, gameToJoin.gameId);
          await gameToJoin?.updateSecondPlayer(user.userId);
          this.pendingGameId = null;
        }
      }

      if (message.type === MOVE) {
        const gameId = message.payload.gameId;
        const game = this.games.find((game) => game.gameId === gameId);
        if (game) {
          game.makeMove(user, message.payload.move);
          if (game.result) {
            this.removeGame(game.gameId);
          }
        }
      }

      if (message.type === JOIN_ROOM) {
        const gameId = message.payload?.gameId;
        if (!gameId) {
          return;
        }

        let availableGame = this.games.find((game) => game.gameId === gameId);
        const gameFromDb = await db.game.findUnique({
          where: { id: gameId },
          include: {
            moves: {
              orderBy: {
                moveNumber: 'asc',
              },
            },
            blackPlayer: true,
            whitePlayer: true,
          },
        });

        if (!gameFromDb) {
          user.socket.send(
            JSON.stringify({
              type: GAME_NOT_FOUND,
            }),
          );
          return;
        }

        if(gameFromDb.status !== GameStatus.IN_PROGRESS) {
          user.socket.send(JSON.stringify({
            type: GAME_ENDED,
            payload: {
              result: gameFromDb.result,
              status: gameFromDb.status,
              moves: gameFromDb.moves,
              blackPlayer: {
                id: gameFromDb.blackPlayer.id,
                name: gameFromDb.blackPlayer.name,
              },
              whitePlayer: {
                id: gameFromDb.whitePlayer.id,
                name: gameFromDb.whitePlayer.name,
              },
            }
          }));
          return;
        }

        if (!availableGame) {
          const game = new Game(
            gameFromDb?.whitePlayerId!,
            gameFromDb?.blackPlayerId!,
            gameFromDb.id,
            gameFromDb.startAt
          );
          game.seedMoves(gameFromDb?.moves || [])
          this.games.push(game);
          availableGame = game;
        }

        console.log(availableGame.getPlayer1TimeConsumed());
        console.log(availableGame.getPlayer2TimeConsumed());

        user.socket.send(
          JSON.stringify({
            type: GAME_JOINED,
            payload: {
              gameId,
              moves: gameFromDb.moves,
              blackPlayer: {
                id: gameFromDb.blackPlayer.id,
                name: gameFromDb.blackPlayer.name,
              },
              whitePlayer: {
                id: gameFromDb.whitePlayer.id,
                name: gameFromDb.whitePlayer.name,
              },
              player1TimeConsumed: availableGame.getPlayer1TimeConsumed(),
              player2TimeConsumed: availableGame.getPlayer2TimeConsumed(),
            },
          }),
        );

        SocketManager.getInstance().addUser(user, gameId);
      }
    });
  }
}

File: apps/ws-rust/src/game/game_manager.rs
use crate::db::{Database, models};
use crate::game::GameState;
use crate::messages::{
    GameJoinedPayload, InitGamePayload, JoinRoomPayload, MessagePayload, MessageType,
};
use crate::socket_manager::SocketManager;
use std::collections::HashMap;

pub struct User {
    pub id: usize,
    pub user_id: String,
    pub sender: Sender,
}

impl User {
    pub fn new(sender: Sender, user_id: String) -> Self {
        User {
            id: rand::thread_rng().gen::<usize>(),
            user_id,
            sender,
        }i
    }
}

pub struct GameManager {
    games: HashMap<String, GameState>,
    pending_game_id: Option<String>,
    database: Database,
}

impl GameManager {
    pub fn new(database: Database) -> Self {
        GameManager {
            games: HashMap::new(),
            pending_game_id: None,
            database,
        }
    }

    pub fn init_game(&mut self, payload: MessagePayload) {
        if let Some(pending_game_id) = self.pending_game_id.take() {
            if let Some(game) = self.games.get_mut(&pending_game_id) {
                if let InitGamePayload::JoinGame(user_id) = payload {
                    if user_id == game.game.white_player.id {
                        SocketManager::broadcast(
                            pending_game_id.clone(),
                            messages::Message {
                                r#type: messages::MessageType::GameAlert,
                                payload: "Trying to Connect with yourself?".to_string(),
                            },
                        );
                        return;
                    }

                    game.game.black_player = Some(models::User {
                        id: user_id.clone(),
                        name: "".to_string(),
                    });

                    self.database.create_game(game.game.clone());

                    SocketManager::broadcast(
                        pending_game_id.clone(),
                        messages::Message {
                            r#type: messages::MessageType::InitGame,
                            payload: InitGamePayload::GameInitialized {
                                game_id: pending_game_id.clone(),
                                white_player: game.game.white_player.clone(),
                                black_player: game.game.black_player.clone(),
                                fen: game.chess.fen().to_string(),
                                moves: vec![],
                            },
                        },
                    );
                }
            }
        } else {
            if let InitGamePayload::CreateGame(user_id) = payload {
                let white_player = models::User {
                    id: user_id.clone(),
                    name: "".to_string(),
                };
                let game = GameState::new(models::Game {
                    id: uuid::Uuid::new_v4().to_string(),
                    white_player,
                    black_player: None,
                    start_time: Utc::now(),
                    moves: vec![],
                    current_fen: "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1"
                        .to_string(),
                    status: GameStatus::InProgress,
                    result: None,
                });

                let game_id = game.game.id.clone();
                self.games.insert(game_id.clone(), game);
                self.pending_game_id = Some(game_id);

                SocketManager::broadcast(
                    game_id,
                    messages::Message {
                        r#type: messages::MessageType::GameAdded,
                        payload: MessagePayload::None,
                    },
                );
            }
        }
    }

    pub fn make_move(&mut self, payload: MessagePayload) {
        if let MessagePayload::MovePayload(move_payload) = payload {
            if let Some(game) = self.games.get_mut(&move_payload.game_id) {
                if let Err(err) = game.make_move(move_payload, &move_payload.user_id) {
                    SocketManager::broadcast(
                        move_payload.game_id.clone(),
                        messages::Message {
                            r#type: messages::MessageType::GameAlert,
                            payload: err,
                        },
                    );
                }
            }
        }
    }

    pub fn join_room(&mut self, payload: JoinRoomPayload) {
        let game_id = payload.game_id.unwrap();

        if let Some(mut game) = self.games.remove(&game_id) {
            SocketManager::broadcast(
                game_id.clone(),
                messages::Message {
                    r#type: messages::MessageType::GameJoined,
                    payload: GameJoinedPayload {
                        game_id: game_id.clone(),
                        moves: game.game.moves.clone(),
                        black_player: game.game.black_player.clone(),
                        white_player: game.game.white_player.clone(),
                        player1_time_consumed: game.player1_time_consumed,
                        player2_time_consumed: game.player2_time_consumed,
                    },
                },
            );

            self.games.insert(game_id, game);
        } else {
            if let Some(game) = self.database.find_game(&game_id) {
                let game_state = GameState::new(game.clone());
                self.games.insert(game_id.clone(), game_state);

                SocketManager::broadcast(
                    game_id.clone(),
                    messages::Message {
                        r#type: messages::MessageType::GameJoined,
                        payload: GameJoinedPayload {
                            game_id: game_id.clone(),
                            moves: game.moves.clone(),
                            black_player: game.black_player.clone(),
                            white_player: game.white_player.clone(),
                            player1_time_consumed: 0,
                            player2_time_consumed: 0,
                        },
                    },
                );
            } else {
                SocketManager::broadcast(
                    game_id,
                    messages::Message {
                        r#type: messages::MessageType::GameNotFound,
                        payload: MessagePayload::None,
                    },
                );
            }
        }
    }
}